client-data/tools/grid/grid.js

```diff
@@ -0,0 +1,86 @@
+/**
+ *                        WHITEBOPHIR
+ *********************************************************
+ * @licstart  The following is the entire license notice for the
+ *  JavaScript code in this page.
+ *
+ * Copyright (C) 2013  Ophir LOJKINE
+ *
+ *
+ * The JavaScript code in this page is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU
+ * General Public License (GNU GPL) as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.  The code is distributed WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
+ *
+ * As additional permission under GNU GPL version 3 section 7, you
+ * may distribute non-source (e.g., minimized or compacted) forms of
+ * that code without the copy of the GNU GPL normally required by
+ * section 4, provided you include this license notice and a URL
+ * through which recipients can access the Corresponding Source.
+ *
+ * @licend
+ */
+
+(function grid() { //Code isolation
+
+    var index = 0; //grid off by default
+    var states = ["none", "url(#grid)", "url(#dots)"];
+
+    function toggleGrid(evt) {
+        index = (index+1) % states.length;
+        gridContainer.setAttributeNS(null, "fill", states[index]);
+    }
+
+    function createSVGElement(name, attrs) {
+        var elem = document.createElementNS("http://www.w3.org/2000/svg", name);
+        Object.keys(attrs).forEach(function(key, i) {
+            elem.setAttributeNS(null, key, attrs[key]);
+        });
+        return elem;
+    }
+
+    function init() {
+        // create patterns
+        // small (inner) grid
+        var smallGrid = createSVGElement("pattern", {id: "smallGrid", width: "30", height: "30", patternUnits: "userSpaceOnUse"});
+        smallGrid.appendChild(
+            createSVGElement("path", {d: "M 30 0 L 0 0 0 30", fill: "none", stroke: "gray", 'stroke-width': "0.5"})
+        );
+        // (outer) grid
+        var grid = createSVGElement("pattern", {id: "grid", width: "300", height: "300", patternUnits: "userSpaceOnUse"});
+        grid.appendChild(createSVGElement("rect", {width: "300", height: "300", fill: "url(#smallGrid)"}));
+        grid.appendChild(
+            createSVGElement("path", {d: "M 300 0 L 0 0 0 300", fill: "none", stroke: "gray", 'stroke-width': "1"})
+        );
+        // dots
+        var dots = createSVGElement("pattern", {id: "dots", width: "30", height: "30", x: "-10", y: "-10", patternUnits: "userSpaceOnUse"});
+        dots.appendChild(createSVGElement("circle", {fill: "gray", cx: "10", cy: "10", r: "2"}));
+
+        var defs = Tools.svg.getElementById("defs");
+        defs.appendChild(smallGrid);
+        defs.appendChild(grid);
+        defs.appendChild(dots);
```

Comment:
Maybe we could define the grid in an external SVG file, to keep the code more readable ?

Reply:
unfortunately IE does not support including external svg with `use` which would be the way to keep the the svg out of the js code. 
However I have just tried testing the site in IE11 and noticed several other problems anyway:
- Promises are not supported (used for the `batchCall` function) for which we could use a polyfill
- multiple svgs for the icons are not rendering correctly (probably need to change the viewBox) which is fixable
- `window.scrollX` and `window.scrollY` don't exist, instead we can use `document.documentElement.scrollLeft` and `document.documentElement.scrollTop`
- only the hand tool is usable at all and even then only after I applied the above fixes.

I just went through the history to see when these were introduced and noticed that the `scrollX` one is from 2013 (5bdf547) and the `Promise` is from over 7 months ago (e03852). I think we should just give up on IE given that it isn't supported at the moment anyhow.

Reply:
Thank you very much for taking the time to test with Internet Explorer and listing issues ! The missing features of Internet Explorer that you list don't seem to be too fundamental to the way WBO works... Do you think we could get a better browser compatibility without having to write too much specialized code just by integrating something like [polyfill-library](https://www.npmjs.com/package/polyfill-library) ? Would you be interested in working on that ? I don't have a windows computer, but I can test things in a VM.

Reply:
I'm investigating the necessary changes on the [bug/ie-fixes](https://github.com/finnboeger/whitebophir/tree/bug/ie-fixes) branch. I have a Windows VM set up as well for that purpose.

Reply:
It looks like it is going to take some deep digging to get it working in IE.
The primary issues are as the following:
- the `removeEventListener` call when switching tools is not removing the listeners which leads to malformed messages. If a user selects any tool and then subsequently any different tool both listeners will be active at the same time leading to the a jumbled mess where one tool will attempt to draw stuff with the message for the other tool.
- IE constantly emits `mouseleave` events for some reason making it nigh impossible to draw anything. This can at least be easily fixed using UA sniffing.

Reply:
Just had another idea and noticed that the issue has nothing to do with the closures but rather Internet Explorer seems to decide on its own whether the `useCapture` option should be set to true. 
By also attempting to remove the listener in capturing mode if the browser is IE we can fix this